core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp

```diff
@@ -0,0 +1,1590 @@
+void OpenACCParallelReduceSum(OpenACCMDRangeBegin<5> const& begin,
+                              OpenACCMDRangeEnd<5> const& end, ValueType& val,
+                              Functor const& functor, int const async_arg) {
+  auto const begin4 = begin[4];
+  auto const end4   = end[4];
+  auto const begin3 = begin[3];
+  auto const end3   = end[3];
+  auto const begin2 = begin[2];
+  auto const end2   = end[2];
+  auto const begin1 = begin[1];
+  auto const end1   = end[1];
+  auto const begin0 = begin[0];
+  auto const end0   = end[0];
+
+  if ((end0 <= begin0) || (end1 <= begin1) || (end2 <= begin2) ||
+      (end3 <= begin3) || (end4 <= begin4)) {
+    return;
+  }
+
+  auto const a_functor(functor);
+  auto l_val = val;
+
+// clang-format off
+#pragma acc parallel loop gang vector collapse(5) reduction(+ : l_val) copyin(a_functor) async(async_arg)
+  // clang-format on
+  for (auto i4 = begin4; i4 < end4; i4++) {
+    for (auto i3 = begin3; i3 < end3; i3++) {
+      for (auto i2 = begin2; i2 < end2; i2++) {
+        for (auto i1 = begin1; i1 < end1; i1++) {
+          for (auto i0 = begin0; i0 < end0; i0++) {
+            a_functor(i0, i1, i2, i3, i4, l_val);
+          }
+        }
+      }
+    }
+  }
+  val = l_val;
+}
+
+template <class ValueType, class Functor>
+void OpenACCParallelReduceSum(OpenACCMDRangeBegin<6> const& begin,
+                              OpenACCMDRangeEnd<6> const& end, ValueType& val,
+                              Functor const& functor, int const async_arg) {
+  auto const begin5 = begin[5];
+  auto const end5   = end[5];
+  auto const begin4 = begin[4];
+  auto const end4   = end[4];
+  auto const begin3 = begin[3];
+  auto const end3   = end[3];
+  auto const begin2 = begin[2];
+  auto const end2   = end[2];
+  auto const begin1 = begin[1];
+  auto const end1   = end[1];
+  auto const begin0 = begin[0];
+  auto const end0   = end[0];
+
+  if ((end0 <= begin0) || (end1 <= begin1) || (end2 <= begin2) ||
+      (end3 <= begin3) || (end4 <= begin4) || (end5 <= begin5)) {
+    return;
+  }
+
+  auto const a_functor(functor);
+  auto l_val = val;
+
+// clang-format off
+#pragma acc parallel loop gang vector collapse(6) reduction(+ : l_val) copyin(a_functor) async(async_arg)
+  // clang-format on
+  for (auto i5 = begin5; i5 < end5; i5++) {
+    for (auto i4 = begin4; i4 < end4; i4++) {
+      for (auto i3 = begin3; i3 < end3; i3++) {
+        for (auto i2 = begin2; i2 < end2; i2++) {
+          for (auto i1 = begin1; i1 < end1; i1++) {
+            for (auto i0 = begin0; i0 < end0; i0++) {
+              a_functor(i0, i1, i2, i3, i4, i5, l_val);
+            }
+          }
+        }
+      }
+    }
+  }
+  val = l_val;
+}
+
+template <class ValueType, class Functor>
+void OpenACCParallelReduceProd(OpenACCMDRangeBegin<2> const& begin,
+                               OpenACCMDRangeEnd<2> const& end, ValueType& val,
+                               Functor const& functor, int const async_arg) {
```

Comment:
This only differs minimally from `OpenACCParallelReduceSum`. Can't you unify the definition for different kinds of reductions using a macro?

Reply:
Refactored similar code patterns using macro functions.